llvm/lib/Target/AMDGPU/SIISelLowering.cpp

```diff
@@ -6899,7 +6899,17 @@ SDValue SITargetLowering::lowerFP_ROUND(SDValue Op, SelectionDAG &DAG) const {
     if (Op.getOpcode() != ISD::FP_ROUND)
       return Op;
 
-    SDValue FpToFp16 = DAG.getNode(ISD::FP_TO_FP16, DL, MVT::i32, Src);
+    if (!Subtarget->has16BitInsts()) {
+      SDValue FpToFp16 = DAG.getNode(ISD::FP_TO_FP16, DL, MVT::i32, Src);
+      SDValue Trunc = DAG.getNode(ISD::TRUNCATE, DL, MVT::i16, FpToFp16);
+      return DAG.getNode(ISD::BITCAST, DL, MVT::f16, Trunc);
+    }
+    if (getTargetMachine().Options.UnsafeFPMath) {
+      SDValue Flags = Op.getOperand(1);
+      SDValue Src32 = DAG.getNode(ISD::FP_ROUND, DL, MVT::f32, Src, Flags);
+      return DAG.getNode(ISD::FP_ROUND, DL, MVT::f16, Src32, Flags);
+    }
```

Comment:
Ought to start using the new fast math flags on fp casts. But I don't see how this unsafe path belongs here, and not in LowerF64ToF16

Reply:
The LowerF64ToF16 should just be the safe path. Probably should change the function name for better clarification

Reply:
Renamed `LowerF64ToF16`  to `LowerF64ToF16Safe`. LowerF64ToF16 is used in target supports f16 and which does not and the unsafe path return node in different types. I think it's a bit ugly to put them in the same utility function and check for targets vs math mode, and cast/trunc.